idea-plugin/src/main/kotlin/arrow/meta/dsl/ide/editor/lineMarker/LineMarkerSyntax.kt

```diff
@@ -11,44 +11,51 @@ import com.intellij.openapi.actionSystem.AnAction
 import com.intellij.openapi.editor.markup.GutterIconRenderer
 import com.intellij.openapi.progress.ProgressManager
 import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiNameIdentifierOwner
+import com.intellij.psi.util.PsiTreeUtil
 import javax.swing.Icon
 
 interface LineMarkerSyntax {
 
   /**
    * This technique adds an LineMarker on the specified PsiElement similar to the Recursive Kotlin Icon [org.jetbrains.kotlin.idea.highlighter.KotlinRecursiveCallLineMarkerProvider]
-   * or Suspended Icon [org.jetbrains.kotlin.idea.highlighter.KotlinSuspendCallLineMarkerProvider]
+   * or Suspended Icon [org.jetbrains.kotlin.idea.highlighter.KotlinSuspendCallLineMarkerProvider].
    * TODO: Add more Techniques such as the one from Elm
    */
   @Suppress("UNCHECKED_CAST")
   fun <A : PsiElement> IdeMetaPlugin.addLineMarkerProvider(
     icon: Icon,
-    matchOn: (PsiElement) -> A?,
+    transform: (PsiElement) -> A?,
     message: (element: A) -> String = Noop.string1(),
     placed: GutterIconRenderer.Alignment = GutterIconRenderer.Alignment.RIGHT
   ): ExtensionPhase =
     addLineMarkerProvider(
-      matchOn,
-      { a ->
-        lineMarkerInfo(icon, a, message as (PsiElement) -> String, placed)
+      transform,
+      {
+        lineMarkerInfo(icon, (it as? PsiNameIdentifierOwner)?.identifyingElement
```

Comment:
Great idea to use the identifying element! That's much more generic.

Perhaps add a note in the comment or rename the method to tell a developer that it's automatically choosing a leaf for the result of `transform()`?

I took another look at the IntelliJ SDK. It seems to be assuming that a line marker is returned for the `PsiElement` which was passed to the line marker provider. But the relation `PsiElement <-> LineMarker` isn't used outside of the marker itself, as far as I can tell. 
So this implementation should work fine. LineMarkers created here might do unexpected things in the future when JetBrains changes the internals. 
2019.3 EAP is still following the same logic here as 2019.2.

Reply:
Yes, the User shouldn't bother with these implementation details. I am still refining the DSL so we can merge your tests without any changes on User code. I saw the Errors, too. I think it tries to associate the passed element to the composite. Which generally leads to `ClassCastExceptions`. 
KtTokens work, but constructors seem to bail.

Thank you for checking in with me. This is all a step by step improvement.